src/content_stream.cpp

```diff
@@ -52,32 +52,41 @@ std::string ContentReader::read() {
   while (matchAny(" \t")) {
   }
   // Parse length
-  size_t len = 0;
+  int len = 0;
   while (true) {
     auto c = matchAny("0123456789");
     if (c == 0) {
       break;
     }
     len *= 10;
-    len += size_t(c) - size_t('0');
+    len += static_cast<int>(c) - static_cast<int>('0');
   }
-  if (len == 0) {
+  if (len <= 0) {
     return "";
   }
   // Expect \r\n\r\n
   if (!match("\r\n\r\n")) {
     return "";
   }
+
   // Read message
-  if (!buffer(len)) {
-    return "";
-  }
-  std::string out;
-  out.reserve(len);
-  for (size_t i = 0; i < len; i++) {
-    out.push_back(static_cast<char>(buf.front()));
+  std::string out(len, 0);
```

Comment:
So the goal of these changes are to avoid the intermediate reading into `buf` before copying to `out`, right?

Do you think we could move this out to a separate change? I think this is a good optimization, but doesn't really belong in a change called `Fix socket receiving error`.

I also have the following thoughts:
* I'm not keen on the `const_cast` on the returned pointer from `std::string::data()`. [Modifying the character array accessed through the const overload of data has undefined behavior](https://en.cppreference.com/w/cpp/string/basic_string/data). Unfortunately the non-const pointer overloads of `std::string::data()` [are missing until C++17](https://en.cppreference.com/w/cpp/string/basic_string/data).
* Perhaps we should have a private `std::vector<uint8_t> ContentReader::readRaw(size_t)` method that takes from `buf` before performing the `reader->read()` loop. This will reduce the bloat of this `ContentReader::read()` method.